src/Cns.ts

```diff
@@ -3,38 +3,61 @@ import {
   NamingServiceName,
   RegistryMap,
   ResolutionResponse,
-  isNullAddress,
-  nodeHash,
   SourceDefinition,
+  isNullAddress,
   NullAddress,
 } from './types';
-import { default as resolverInterface } from './cns/contract/resolver';
-import { default as cnsInterface } from './cns/contract/registry';
-import ResolutionError from './errors/resolutionError';
-import { ResolutionErrorCode } from './errors/resolutionError';
+import { default as proxyReaderAbi } from './cns/contract/proxyReader';
+import ResolutionError, { ResolutionErrorCode } from './errors/resolutionError';
+import ICnsReader, { Data } from './cns/ICnsReader';
+import CnsProxyReader from './cns/CnsProxyReader';
+import CnsRegistryReader from './cns/CnsRegistryReader';
 import Contract from './utils/contract';
 
 /** @internal */
 export default class Cns extends EthereumNamingService {
   readonly registryAddress?: string;
   /** @internal */
   readonly RegistryMap: RegistryMap = {
-    mainnet: '0xD1E5b0FF1287aA9f9A268759062E4Ab08b9Dacbe',
-    kovan: '0x22c2738cdA28C5598b1a68Fb1C89567c2364936F', // for internal testing
+    mainnet: '0x7ea9ee21077f84339eda9c80048ec6db678642b1',
+    kovan: '0xcf4318918fd18aca9bdc11445c01fbada4b448e3', // for internal testing
   };
 
+  readonly source: SourceDefinition;
+  readonly contract: Contract;
+
+  reader: ICnsReader;
+
   constructor(source: SourceDefinition = {}) {
     super(source, NamingServiceName.CNS);
-    source = this.normalizeSource(source);
+
     this.registryAddress = source.registry ?
       source.registry :
       this.RegistryMap[this.network];
-    if (this.registryAddress) {
-      this.registryContract = this.buildContract(
-        cnsInterface,
-        this.registryAddress,
-      );
+    this.source = { ...this.normalizeSource(source), registry: this.registryAddress };
+    this.contract = this.buildContract(proxyReaderAbi, this.registryAddress);
+  }
+
+  async getReader(): Promise<ICnsReader> {
+    if (!this.reader) {
+      this.reader = await this.isDataReaderSupported() ?
+        new CnsProxyReader(this.source) :
+        new CnsRegistryReader(this.source);
     }
+    return this.reader;
+  }
+
+  protected async isDataReaderSupported(): Promise<boolean> {
+    try {
+      const [isDataReaderSupported] = await this.contract.call('supportsInterface', ['0x6eabca0d']);
+      if (!isDataReaderSupported) {
+        throw new Error('Not supported DataReader');
+      }
+
+      return true;
+    } catch { }
```

Comment:
One of the reasons why the above call can error out is if the blockchain naming service is down. 
Like no internet connection. 
For example in FetchProvider the code is 
```
if (error instanceof FetchError) {
        throw new ResolutionError(ResolutionErrorCode.NamingServiceDown, {
          method: this.name,
        });
      } else throw error;
```
Do we think it is a good idea to silence the error? Other providers have some errors as well which is I/O related.
Maybe it is better to just return the boolean and let all other errors bubble up?

Reply:
@JohnnyJumper It can be in order to keep consistency, but we can't handle all scenarios. I would prefer to receive I/O error, instead of `NamingServiceDown`, the client may have not internet, but we send `NamingServiceDown`, which is not right.
In the particular case we shall not throw an error because it may be caused be unsupporting the method by smart contract. It is not critical, we instantiate different contract. The function is internal, it will be always executed before `getter` functions and those functions will throw I/O error for sure.

Reply:
I agree about the point with IO errors, though at the same time I'm thinking about consistency with other methods. Which one is worse: having inconsistent error handling design or having ambiguity in error types? I don't have a strong opinion about this myself, so opening a discussion